hoomd/md/Thermostat.h

```diff
@@ -0,0 +1,268 @@
+//
+// Created by girard01 on 11/9/22.
+//
+
+#ifndef HOOMD_THERMOSTAT_H
+#define HOOMD_THERMOSTAT_H
+
+#include <array>
+#include <hoomd/HOOMDMath.h>
+#include <hoomd/ParticleGroup.h>
+#include <hoomd/md/ComputeThermo.h>
+#include <hoomd/Variant.h>
+#include "hoomd/RNGIdentifiers.h"
+#include "hoomd/RandomNumbers.h"
+
+namespace hoomd::md {
+    class Thermostat {
+    public:
+        Thermostat(std::shared_ptr<Variant> T) :
+        m_T(T)
+        {}
+
+        virtual ~Thermostat() {}
+
+        virtual std::array<Scalar, 2> getRescalingFactorsOne(uint64_t timestep, Scalar deltaT) {
+            return {Scalar(1.0), Scalar(1.0)};
+        }
+
+        virtual std::array<Scalar, 2> getRescalingFactorsTwo(uint64_t timestep, Scalar deltaT) {
+            return {Scalar(1.0), Scalar(1.0)};
+        }
+
+        virtual void advanceThermostat(uint64_t timestep, Scalar deltaT, bool aniso) {}
+
+        virtual void thermalizeThermostat() {}
+
+        virtual Scalar getThermostatEnergy(uint64_t timestep) {return Scalar(0.0);}
+
+        std::shared_ptr<Variant> getT(){
+            return m_T;
+        }
+
+        void setT(std::shared_ptr<Variant> T){
+            m_T = T;
+        }
+
+        Scalar getTemperature(uint64_t timestep){
+            return (*m_T)(timestep);
+        }
+
+        void attach(std::shared_ptr<ParticleGroup> group, std::shared_ptr<ComputeThermo> thermo, uint16_t seed){
+            if(is_attached)
+                throw std::runtime_error("Thermostats can only be attached to a single integrator");
```

Comment:
Is there a reason we cannot do this upon construction?

Reply:
I originally suggested this approach when I thought we would use the Variant/Trigger model and users would instantiate C++ objects directly. With the `_HOOMDBaseObject` approach, these can likely be moved back to the constructor. Sorry for the bad advice, @MartinGirard.

Reply:
This has been moved back to the ctor, but required a bit more tracking for the group / compute objects on the thermostat python side.